File: pset6/vigenere/vigenere.py
import sys
from sys import argv
from cs50 import get_string

def main():
    #Check key
    if len(argv) != 2:
        print("Usage: ./vigenere key")
        return 1

    for char in argv[1]:
        if not char.isalpha():
            print("Usage: ./vigenere key should be letters")
            return 1

    key = argv[1]
    #Ask for the input
    plaintxt = get_string("Write plaintxt: ")
    #getting the shift key
    shiftKey = getShiftKey(key)
    #Ciphering
    Cipher(shiftKey, plaintxt)

def Cipher(shiftKey, plaintxt):
    ciphertxt = ""
    c = 0
    for i in range(len(plaintxt)):
        #handle non-char
        if not plaintxt[i].isalpha():
            ciphertxt += (plaintxt[i])
            c += 1
            continue
        key = shiftKey[(i - c) % (len(shiftKey))]
        if plaintxt[i].isupper():
            ciphertxt += chr((((ord(plaintxt[i]) - 65) + key) % 26) + 65)
        else:
            ciphertxt += chr((((ord(plaintxt[i]) - 97) + key) % 26) + 97)
    print(f"ciphertext: {ciphertxt}")


def getShiftKey(key):
    shiftKey = []
    for char in key:
        if ord(char) >= 97:
            shiftKey.append(ord(char) - 97)
        else:
            shiftKey.append(ord(char) - 65)
    return shiftKey




if __name__ == '__main__':
    sys.exit(main())

File: pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

int shift(int key);
int main(int argc, string argv[])
{
    if (argc == 2)
    {
        string key = argv[1];
        //printf("Strlen for key: %lu", strlen(key));
        int Shift[strlen(key)];
        //check terminal-command argument
        for (int i = 0; i < strlen(key); i++)
        {
            if (key[i] < 65)
            {
                printf("Usage: ./caesar key\n");  
                return 1; 
            }
            
            else 
            {
                Shift[i] = shift(key[i]);                
            }
        }
        //Ciphering
        string plaintxt = get_string("Write plaintxt: ");
        int c = 0;
        for (int i = 0; i < strlen(plaintxt); i++)
        {
            if (('A' <= plaintxt[i] && plaintxt[i] <= 'Z'))
            {
                plaintxt[i] -= 65;
                plaintxt[i] = (plaintxt[i] + Shift[(i - c) % strlen(key)]) % 26 + 65;
            }

            else if (('a' <= plaintxt[i] && plaintxt[i] <= 'z'))
            {       
                plaintxt[i] -= 97;
                plaintxt[i] = (plaintxt[i] + Shift[(i - c) % strlen(key)]) % 26 + 97;
            }  
            
            else
            {
                c++;                
            }
            
        }
        printf("ciphertext: %s\n", plaintxt);
    }
    
    else
    {
        printf("Usage: ./caesar key\n");  
        return 1;
    }  
}

int shift(int key)
{
    if (key >= 'A' && key <= 'Z')
    {
        return key - 65;
    }
    else if (key >= 'a' && key <= 'z')
    {
        return key - 97;
    }
    else
    {
        printf("Can't calculate the shift for %c \n", key);
        return 1;
    }
}
